### PR TITLE
fix: assertThatBodyParserHasBeenUsedForExpressLikeRequest

### PR DESCRIPTION
**Pull request by bhumilsarvaiya**
## Summary of change

- added `fromNextJS` property for next js request

## Related issues

-   #313

### Test

- [x] disable nextjs parser and verify that the json parser deosn't get stuck and also parses the request body correctly
- [x] disable nextjs praser, modify the code to disable json parser. Call a POST API with correct request body and the request should fail
- [x] send empty json as request body and the request should not get stuck during the json parsing part
- [x] send no request body and the request should not get stuck during the json parsing part
- [x] send correct request body and the API should yield correct response

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [unreleased]
 
+## [9.2.2] - 2022-05-24
+
+### Fixes:
+
+-   Calling the setImmediate function inside assertThatBodyParserHasBeenUsedForExpressLikeRequest only if the function is getting executed in NextJS env
```
Comment (rishabhpoddar) [inline CHANGELOG.md:14]: link to github issue

File: lib/ts/framework/utils.ts [hunk 2]
```diff
@@ -136,17 +136,19 @@ export async function assertThatBodyParserHasBeenUsedForExpressLikeRequest(
...
+                     * where the json parser would not resolve and thus the request
+                     * just hangs forever. Next.JS does json parsing on its own.
+                     */
+                    setImmediate(() => {
+                        if (!resolvedCalled) {
+                            resolvedCalled = true;
+                            resolve(undefined);
+                        }
+                    });
+                }
```
Comment (rishabhpoddar) [inline lib/ts/framework/utils.ts:151]: If it's next js, we don't need to parse anything right? So you can remove the setImmediate and the jsonParser if this if statmeent is true??

File: lib/ts/nextjs.ts [hunk 3]
```diff
@@ -39,6 +39,7 @@ export default class NextJS {
         response: any
     ): Promise<T> {
         return new Promise<T>(async (resolve: any, reject: any) => {
+            request.fromNextJS = true;
```
Comment (rishabhpoddar) [inline lib/ts/nextjs.ts:42]: I prefer a name like `request.__supertokensFromNextJS` just so it's clear this is a var from us

Comment (rishabhpoddar) [inline lib/ts/nextjs.ts:42]: You should add a test that if this wrapper is used, then the request has this variable.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/framework/utils.ts (hunk 2 above)
Comment (bhumilsarvaiya) [reply lib/ts/framework/utils.ts]: nextjs do allow users to disable the default parser. In that case, we should be parsing the request body

File: lib/ts/framework/utils.ts [hunk 4]
```diff
@@ -137,16 +137,22 @@ export async function assertThatBodyParserHasBeenUsedForExpressLikeRequest(
             let err = await new Promise((resolve) => {
                 let resolvedCalled = false;
                 /**
-                 * the setImmediate here is to counter the next.js issue
-                 * where the json parser would not resolve and thus the request
-                 * just hangs forever. Next.JS does json parsing on its own.
+                 * Nextjs allow users to disable the default parser.
+                 * To handle that scenario, we are still parsing the request body
                  */
```
Comment (rishabhpoddar) [inline lib/ts/framework/utils.ts:142]: Need to test that if user has disabled bodyParser in NextJS, our parser runs and it doesn't get "stuck"

File: test/nextjs.test.js [hunk 5]
```diff
@@ -404,6 +404,100 @@ describe(`NextJS Middleware Test: ${printPath("[test/nextjs.test.js]")}`, functi
...
+                return done();
+            });
+
+            superTokensNextWrapper(
+                async (next) => {
+                    return middleware()(request, response, next);
+                },
+                request,
+                response
+            );
```
Comment (rishabhpoddar) [inline test/nextjs.test.js:497]: how is this test passing if the middleware is added after the request is being made?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**